### PR TITLE
PublicDashboards: Resolve interval for public dashboard data source

### LINKED ISSUE #55304: Public Dashboards: Datapoints granularity is different on public dashboard
**What happened**:

The datapoints are different for public dashboard.

On the non-public dashboard the points are displayed every 15 seconds, while on public dashboard they are displayed every minute.

Default dashboard:
Default dashboard

Public dashboard:
Public dashboard

**What you expected to happen**:

I expected the data points to be displayed at the same frequency.

**How to reproduce it (as minimally and precisely as possible)**:

Pre-requirement: 
A Grafana instance with the gdev-prometheus datasource installed (`make devenv sources=prometheus`) and public dashboards enabled on the .ini (`[feature_toggles]publicDashboards = true`)

1. Create a dashboard
2. Add a panel with the query `go_goroutines{instance="host.docker.internal:3000"}`
3. Set the time range to the last 15m and save the dashboard
4. Click the share button, go to public dashboards, make it public
5. Access the link and compare both

**Anything else we need to know?**:

Comparing the queries on the front-end I noticed the following discrepancy:

Private dashboard:
```
intervalMs: 15000
maxDataPoints: 672
```

Public dashboard:
```
intervalMs: 60000
maxDataPoints: 700
```

**Environment**:
- Grafana version: main(5c66c412200b8f756851ef84da0a4130b9bbc16a)
- Data source type & version: Prometheus (prom/prometheus:latest) from devenv
- OS Grafana is installed on: macOS Monterey 12.6
- User OS & Browser: Chrome Version 105.0.5195.125 (Official Build) (arm64) on macOS Monterey 12.6
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by guicaulada**
**What this PR does / why we need it**:

This pull request adds a call to `getDatasourceSrv().get()` on the `PanelQueryRunner` `getDataSource` function to retrieve the datasource data before passing it as an argument to `PublicDashboardDataSource`, this way the `PublicDashboardDataSource` can call the newly created function `resolveInterval` to determine the interval of the datasource, fixing the issue with data granularity discrepancies between private and public dashboards.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/55304

**Special notes for your reviewer**:

Look issue for exmaple pictures.

### CODE REVIEW

**Comment by guicaulada** [comment]
This PR removes some `any` type assertions and uses the correct types to fix linter issues and make things even "betterer" 😅

File: public/app/features/query/state/PanelQueryRunner.ts [hunk 1]
```diff
@@ -370,13 +370,14 @@ async function getDataSource(
...
 ): Promise<DataSourceApi> {
-  if (publicDashboardAccessToken) {
-    return new PublicDashboardDataSource(datasource);
+  if (!publicDashboardAccessToken && datasource && (datasource as DataSourceApi).query) {
+    return datasource as DataSourceApi;
   }
 
-  if (datasource && (datasource as any).query) {
-    return datasource as DataSourceApi;
+  const ds = await getDatasourceSrv().get(datasource, scopedVars);
```
Comment (owensmallwood) [inline public/app/features/query/state/PanelQueryRunner.ts:377]: I could have sworn you can't use getDatasourceSrv() while unauthed. But I tested it out and it works fine. I was smokin beers I guess 🤷‍♂️

**Review by owensmallwood — APPROVED** [review]
Nice fix 👍 🦖 💯

**Review by joshhunt — COMMENTED** [review]
Our goal with the betterer is to eventually remove it and graduate all it's lint rules just into eslint directly, do we don't add any new exceptions into it. 

As an option of last resort, we can use `// eslint-ignore` comments so you must 'pay the cost' of having the smell directly in the code. Or, we can just write our code to not use them ✨

File: public/app/features/dashboard/services/PublicDashboardDataSource.ts [hunk 2]
```diff
@@ -54,10 +56,18 @@ export class PublicDashboardDataSource extends DataSourceApi<any> {
     return datasource?.uid === MIXED_DATASOURCE_NAME;
   }
 
+  private static resolveInterval(datasource: DataSourceRef | string | DataSourceApi | null): string {
+    if (typeof datasource === 'string' || datasource === null) {
+      return DEFAULT_INTERVAL;
+    }
+
+    return (datasource as DataSourceApi)?.interval ?? DEFAULT_INTERVAL;
+  }
```
[the remark this answers is not in the document]
Comment (guicaulada) [reply public/app/features/dashboard/services/PublicDashboardDataSource.ts:65]: Nice! This does looks much better! Thank you!

File: public/app/features/query/state/PanelQueryRunner.ts [hunk 3]
```diff
@@ -370,13 +370,14 @@ async function getDataSource(
...
+  if (publicDashboardAccessToken) {
+    return new PublicDashboardDataSource(ds);
   }
 
-  return await getDatasourceSrv().get(datasource as string, scopedVars);
+  return ds;
```
[the remark this answers is not in the document]
Comment (guicaulada) [reply public/app/features/query/state/PanelQueryRunner.ts:382]: Same as above! This looks much better! Thank you!

[the remark this answers is not in the document]
Comment (guicaulada) [reply public/app/features/query/state/PanelQueryRunner.ts]: Done on https://github.com/grafana/grafana/pull/55489/commits/b8fd65bb9529f531a58c3ce01c5fa15a934a79cb

File: public/app/features/dashboard/services/PublicDashboardDataSource.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (guicaulada) [reply public/app/features/dashboard/services/PublicDashboardDataSource.ts]: Done on https://github.com/grafana/grafana/pull/55489/commits/b8fd65bb9529f531a58c3ce01c5fa15a934a79cb

File: public/app/features/query/state/PanelQueryRunner.ts (hunk 1 above)
Comment (guicaulada) [reply public/app/features/query/state/PanelQueryRunner.ts:377]: You weren't smoking beers! Tested it on an anonymous window and the page doesn't load.. so turns out we are not allowed to call this... I'm wondering how can we get the interval then.

Comment (owensmallwood) [reply public/app/features/query/state/PanelQueryRunner.ts:377]: Weird, I just did the same thing and it works fine. I'll rebuild stuff and see if that changes it

Comment (guicaulada) [reply public/app/features/query/state/PanelQueryRunner.ts:377]: This just got a lot more complicated.

Comment (guicaulada) [reply public/app/features/query/state/PanelQueryRunner.ts:377]: Fixed on https://github.com/grafana/grafana/pull/55489/commits/f31996dfab174d8a008e9f22c10c671ec961b4e1

**Pull request merged**